File: frontend/src/infrastructure/bridge/bridgeError.ts
import {
  type AppError,
  type AppErrorFieldError,
  bridgeCallFailed,
} from "../../application/errors/appError";

/**
 * The marker the backend bridge puts in front of the JSON error model. Wails 2
 * carries only `error.Error()` back from a bound method, so the structured
 * contract travels inside that one string; this prefix is the same constant the
 * Go side writes and nothing here guesses at an undocumented serialization.
 */
export const bridgeErrorPrefix = "saveforge-error:";

/**
 * Recovers the backend's structured failure from a caught rejection.
 *
 * The validation is strict on purpose: an envelope that is absent, truncated,
 * not valid JSON, not an object, or missing a member of the contract is not
 * "partly usable". It is reduced to `bridge_call_failed`, because the
 * alternative is showing the user a failure the frontend invented from a
 * fragment.
 */
export function parseBridgeError(reason: unknown): AppError {
  const message = reason instanceof Error ? reason.message : null;
  if (message === null || !message.startsWith(bridgeErrorPrefix)) {
    return bridgeCallFailed();
  }
  const payload = message.slice(bridgeErrorPrefix.length);
  if (payload === "") {
    return bridgeCallFailed();
  }

  let decoded: unknown;
  try {
    decoded = JSON.parse(payload);
  } catch {
    return bridgeCallFailed();
  }
  if (typeof decoded !== "object" || decoded === null || Array.isArray(decoded)) {
    return bridgeCallFailed();
  }

  const raw = decoded as Record<string, unknown>;
  const code = stringMember(raw.code);
  const errorMessage = stringMember(raw.message);
  const severity = stringMember(raw.severity);
  const stage = stringMember(raw.stage);
  const diagnosticID = stringMember(raw.diagnosticID);
  if (
    code === null ||
    errorMessage === null ||
    severity === null ||
    stage === null ||
    diagnosticID === null ||
    typeof raw.retryable !== "boolean"
  ) {
    return bridgeCallFailed();
  }
  const params = stringMap(raw.params);
  const fieldErrors = fieldErrorList(raw.fieldErrors);
  if (params === null || fieldErrors === null) {
    return bridgeCallFailed();
  }
  const currentRevision = optionalStringMember(raw.currentRevision);
  if (currentRevision === undefined) {
    return bridgeCallFailed();
  }

  return {
    code,
    message: errorMessage,
    params,
    severity,
    stage,
    retryable: raw.retryable,
    fieldErrors,
    currentRevision,
    diagnosticID,
  };
}

function stringMember(value: unknown): string | null {
  return typeof value === "string" && value !== "" ? value : null;
}

/** Returns the string, `null` for an absent member, `undefined` for a wrong type. */
function optionalStringMember(value: unknown): string | null | undefined {
  if (value === undefined || value === null) {
    return null;
  }
  return typeof value === "string" ? value : undefined;
}

function stringMap(value: unknown): Record<string, string> | null {
  if (value === undefined || value === null) {
    return {};
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  const result: Record<string, string> = {};
  for (const [key, member] of Object.entries(value as Record<string, unknown>)) {
    if (typeof member !== "string") {
      return null;
    }
    result[key] = member;
  }
  return result;
}

function fieldErrorList(value: unknown): AppErrorFieldError[] | null {
  if (value === undefined || value === null) {
    return [];
  }
  if (!Array.isArray(value)) {
    return null;
  }
  const result: AppErrorFieldError[] = [];
  for (const entry of value) {
    if (typeof entry !== "object" || entry === null || Array.isArray(entry)) {
      return null;
    }
    const raw = entry as Record<string, unknown>;
    const field = stringMember(raw.field);
    const code = stringMember(raw.code);
    const message = stringMember(raw.message);
    if (field === null || code === null || message === null) {
      return null;
    }
    result.push({ field, code, message });
  }
  return result;
}

File: internal/desktop/bridge_error.go
package desktop

import (
	"encoding/json"
	"errors"
	"log"
	"strings"

	"github.com/oisis/EldenRing-SaveForge/backend/apperror"
)

// This file is the error boundary of the desktop bridge.
//
// Wails 2.15 carries exactly one thing back from a failing bound method: the
// string of the returned error. internal/frontend/dispatcher/calls.go assigns
// `callbackMessage.Err = err.Error()`, and the desktop runtime turns that
// string into `new Error(message)` before rejecting the call. There is no
// structured error channel to opt into, so a structured contract has to be
// carried inside that one string.
//
// The envelope below is therefore deliberate and explicit rather than a guess
// about undocumented serialization: every bridge failure is one line consisting
// of a fixed marker and the JSON of the shared public error model. The frontend
// adapter parses exactly that shape, validates it, and reduces anything else —
// a runtime failure, a Wails timeout, a truncated payload — to its safe
// bridge_call_failed code.
//
// Domain results keep their generated types; only the error side is encoded.

// BridgeErrorPrefix marks a bridge error string that carries the public error
// model as JSON. The frontend matches this exact prefix, so it is a shared
// constant and never a literal on either side.
const BridgeErrorPrefix = "saveforge-error:"

// bridgeError converts one endpoint failure into the transportable envelope.
// The private cause is logged here with the same diagnosticID the frontend
// receives and never becomes part of the string that leaves the process.
func bridgeError(err error) error {
	if err == nil {
		return nil
	}
	public := apperror.From(err)
	if cause := public.Cause(); cause != nil {
		log.Printf("%s %s: %v", public.DiagnosticID, public.Code, cause)
	} else {
		log.Printf("%s %s", public.DiagnosticID, public.Code)
	}
	encoded, marshalErr := json.Marshal(public)
	if marshalErr != nil {
		// Fail closed: an envelope that cannot be built carries the marker and no
		// payload, which the frontend reduces to bridge_call_failed. It never
		// falls back to the raw error text.
		log.Printf("%s: cannot encode the bridge error envelope: %v",
			public.DiagnosticID, marshalErr)
		return errors.New(BridgeErrorPrefix)
	}
	return errors.New(BridgeErrorPrefix + string(encoded))
}

// bridged wraps the result of one endpoint call for the bridge: a success is
// returned unchanged, and a failure is replaced by the envelope together with
// the zero result, so a partial value can never travel beside an error.
//
// ponytail: one generic wrapper instead of an if in every bound method. It is
// the whole error boundary; there is no interceptor chain and no middleware.
func bridged[T any](value T, err error) (T, error) {
	if err != nil {
		var zero T
		return zero, bridgeError(err)
	}
	return value, nil
}

// DecodeBridgeError parses an envelope produced by bridgeError. It exists for
// the bridge's own tests and mirrors exactly what the frontend adapter does: an
// unmarked, truncated or malformed string is not an application error.
func DecodeBridgeError(message string) (*apperror.Error, bool) {
	payload, marked := strings.CutPrefix(message, BridgeErrorPrefix)
	if !marked || payload == "" {
		return nil, false
	}
	var required struct {
		Code         string `json:"code"`
		Message      string `json:"message"`
		Severity     string `json:"severity"`
		Stage        string `json:"stage"`
		Retryable    *bool  `json:"retryable"`
		DiagnosticID string `json:"diagnosticID"`
	}
	if err := json.Unmarshal([]byte(payload), &required); err != nil {
		return nil, false
	}
	if required.Code == "" || required.Message == "" || required.Severity == "" ||
		required.Stage == "" || required.Retryable == nil || required.DiagnosticID == "" {
		return nil, false
	}
	var public apperror.Error
	if err := json.Unmarshal([]byte(payload), &public); err != nil {
		return nil, false
	}
	return &public, true
}
